chainbus/bus_test.go

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2018 The CovenantSQL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package chainbus
+
+import (
+	"sync/atomic"
+	"testing"
+	"time"
+
+	bi "github.com/CovenantSQL/CovenantSQL/blockproducer/interfaces"
+)
+
+func TestNew(t *testing.T) {
+	bus := New()
+	if bus == nil {
+		t.Log("New EventBus not created!")
+		t.Fail()
+	}
+}
+
+func TestHasCallback(t *testing.T) {
+	bus := New()
+	bus.Subscribe(bi.TransactionType(1), func() {})
+	if bus.HasCallback(bi.TransactionType(2)) {
+		t.Fail()
+	}
+	if !bus.HasCallback(bi.TransactionType(1)) {
+		t.Fail()
+	}
+}
+
+func TestSubscribe(t *testing.T) {
+	bus := New()
+	if bus.Subscribe(bi.TransactionType(1), func() {}) != nil {
+		t.Fail()
+	}
+	if bus.Subscribe(bi.TransactionType(1), "String") == nil {
+		t.Fail()
+	}
+}
+
+func TestSubscribeOnce(t *testing.T) {
+	bus := New()
+	if bus.SubscribeOnce(bi.TransactionType(1), func() {}) != nil {
+		t.Fail()
+	}
+	if bus.SubscribeOnce(bi.TransactionType(1), "String") == nil {
+		t.Fail()
+	}
+}
+
+func TestSubscribeOnceAndManySubscribe(t *testing.T) {
+	bus := New()
+	event := bi.TransactionType(1)
+	flag := 0
+	fn := func() { flag += 1 }
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
should replace flag += 1 with flag++